From eb42cc803f567a00ae1460c641445f2938ab9d10 Mon Sep 17 00:00:00 2001
From: Robert Dabrowski <robertx.dabrowski@intel.com>
Date: Mon, 24 Feb 2020 09:54:22 +0100
Subject: [PATCH] PONRTSYS-6739: dp: sleeping function called from invalid
 context

- memory alloc flag GFP_KERNEL change to GFP_ATOMIC
- fix compilation warning
---
 .../net/datapath/mib_counters/dp_mib_counters.c    | 23 +++++++++++++---------
 1 file changed, 14 insertions(+), 9 deletions(-)

diff --git a/drivers/net/datapath/mib_counters/dp_mib_counters.c b/drivers/net/datapath/mib_counters/dp_mib_counters.c
index e82718899f8e..06fb5b7216f0 100644
--- a/drivers/net/datapath/mib_counters/dp_mib_counters.c
+++ b/drivers/net/datapath/mib_counters/dp_mib_counters.c
@@ -164,7 +164,7 @@ s32 add_dev_to_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 
 	dev_cnt = devm_kzalloc(&g_dp_dev->dev,
 			       sizeof(struct mib_counters),
-			       GFP_KERNEL);
+			       GFP_ATOMIC);
 	if (!dev_cnt)
 		return -ENOMEM;
 
@@ -208,7 +208,7 @@ s32 rm_dev_from_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 	port_info = get_dp_port_info(inst, port_id);
 
 	DP_DEBUG(DP_DBG_FLAG_MIB,
-		 "MIB CNT -- looking for item to delete: name = %s, port_id = %d, sub_if = %d, inst = %d\n",
+		 "MIB CNT - looking for item to delete: name = %s, port_id = %d, sub_if = %d, inst = %d\n",
 		 dev->name, port_id, sub_if, inst);
 
 	sub_if_id_grp = GET_VAP(sub_if, port_info->vap_offset,
@@ -218,20 +218,25 @@ s32 rm_dev_from_list(struct net_device *dev, s32 port_id, s32 sub_if, s32 inst)
 	list_for_each_safe(p, q, &registered_devices_list) {
 		mib_cnt = list_entry(p, struct mib_counters, dev_list);
 
-		DP_DEBUG(DP_DBG_FLAG_MIB,
-			 "MIB CNT --- list item: name = %s, port_id = %d, sub_if_id_grp = %d, inst = %d ",
-			 mib_cnt->dev->name, mib_cnt->port_id, mib_cnt->sub_if_id_grp, mib_cnt->inst);
-
 		if ((mib_cnt->port_id == port_id) &&
 		    (mib_cnt->sub_if_id_grp == sub_if_id_grp) &&
 		    (mib_cnt->inst == inst) &&
 		    (!strcmp(mib_cnt->dev->name, dev->name))) {
-			DP_DEBUG(DP_DBG_FLAG_MIB, "REMOVED\n",
-				 dev->name, port_id, sub_if_id_grp);
+			DP_DEBUG(DP_DBG_FLAG_MIB,
+				 "MIB CNT - list item: name = %s, port_id = %d, sub_if_id_grp = %d, inst = %d - removed",
+				 mib_cnt->dev->name,
+				 mib_cnt->port_id,
+				 mib_cnt->sub_if_id_grp,
+				 mib_cnt->inst);
 			list_del(p);
 			devm_kfree(&g_dp_dev->dev, mib_cnt);
 		} else {
-			DP_DEBUG(DP_DBG_FLAG_MIB, "doesn't match\n");
+			DP_DEBUG(DP_DBG_FLAG_MIB,
+				 "MIB CNT - list item: name = %s, port_id = %d, sub_if_id_grp = %d, inst = %d - doesn't match",
+				 mib_cnt->dev->name,
+				 mib_cnt->port_id,
+				 mib_cnt->sub_if_id_grp,
+				 mib_cnt->inst);
 		}
 	}
 	write_unlock(&dev_list_lock);
